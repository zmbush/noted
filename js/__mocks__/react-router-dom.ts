// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
// eslint-disable-next-line import/no-import-module-exports
import type { NavigateFunction, To, NavigateOptions } from 'react-router-dom';

const ReactRouterDom = jest.requireActual('react-router-dom');

const useNavigate = (): NavigateFunction => {
  try {
    return ReactRouterDom.useNavigate();
  } catch {
    return (_to: To | number, _options?: NavigateOptions) => {};
  }
};

module.exports = { ...ReactRouterDom, useNavigate };
