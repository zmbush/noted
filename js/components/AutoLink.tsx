// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

import * as React from 'react';
import { Link } from 'react-router-dom';
import { LinkIdMap } from 'data/selectors';

type LinkProps = {
  text: string;
  ids: Set<number>;
};

const LinkedText = ({ ids, text }: LinkProps) => {
  if (ids.size === 1) {
    const id = ids.values().next().value;
    return <Link to={`/note/${id}`}>{text}</Link>;
  }
  if (ids.size > 1) {
    const theseIds = Array.from(ids.values()).join(',');
    return <Link to={`/disambiguation/${theseIds}`}>{text}</Link>;
  }
  return <>text</>;
};

type Props = {
  titles: LinkIdMap;
  children: string;
};

export default class AutoLink extends React.Component<Props> {
  render() {
    const { children, titles } = this.props;
    let body: any[] = [children];
    let keyIx = 1;

    titles.forEach((value, key) => {
      let newBody: any[] = [];
      body.forEach(part => {
        if (typeof part === 'string' || part instanceof String) {
          const elements = part.split(new RegExp(key, 'i')).reduce((r, a, ix, arr) => {
            r.push(a);
            if (ix + 1 < arr.length) {
              r.push(<LinkedText key={`linked-text-${keyIx}`} text={key} ids={value} />);
              keyIx += 1;
            }
            return r;
          }, []);
          newBody = newBody.concat(elements);
        } else {
          newBody.push(part);
        }
      });
      body = newBody;
    });

    return body;
  }
}
