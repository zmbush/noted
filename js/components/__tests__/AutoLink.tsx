// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

import * as React from 'react';
import { shallow } from 'enzyme';

import AutoLink, { LinkedText } from '../AutoLink';
import { Link } from 'react-router-dom';

describe('<AutoLink />', () => {
  test('works with no titles', () => {
    const wrapper = shallow(<AutoLink titles={new Map()}>Test</AutoLink>);
    expect(wrapper.contains('Test')).toBeTruthy();
  });

  test('works with some titles', () => {
    const titles = new Map();
    titles.set('Goat', new Set([1]));
    titles.set('Boat', new Set([1, 2]));

    const wrapper = shallow(
      <AutoLink titles={titles}>
        This goat is here. It also has boats galore.
      </AutoLink>
    );

    expect(
      wrapper.contains(<LinkedText ids={new Set([1])} text='Goat' />)
    ).toBeTruthy();
    expect(
      wrapper.contains(<LinkedText ids={new Set([1, 2])} text='Boat' />)
    ).toBeTruthy();
  });
});

describe('<LinkedText />', () => {
  test('works with no ids', () => {
    const wrapper = shallow(<LinkedText ids={new Set()} text='Link' />);
    expect(wrapper.contains('Link')).toBeTruthy();
  });

  test('works with 1 id', () => {
    const wrapper = shallow(<LinkedText ids={new Set([1])} text='Link' />);
    expect(wrapper.contains(<Link to='/note/1'>Link</Link>)).toBeTruthy();
  });

  test('works with more than 1 id', () => {
    const wrapper = shallow(
      <LinkedText ids={new Set([1, 2, 3])} text='Link' />
    );
    expect(
      wrapper.contains(<Link to='/disambiguation/1,2,3'>Link</Link>)
    ).toBeTruthy();
  });
});
