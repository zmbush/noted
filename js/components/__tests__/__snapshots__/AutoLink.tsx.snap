// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutoLink /> matches snapshot 1`] = `
Array [
  "This ",
  <LinkedText
    ids={
      Set {
        1,
      }
    }
    key="linked-text-1"
    text="Goat"
  />,
  " is here. It also has ",
  <LinkedText
    ids={
      Set {
        1,
        2,
      }
    }
    key="linked-text-2"
    text="Boat"
  />,
  "s galore.",
]
`;

exports[`<LinkedText /> matches snapshot 1`] = `"Link"`;

exports[`<LinkedText /> matches snapshot 2`] = `
<Link
  to="/note/1"
>
  Link
</Link>
`;

exports[`<LinkedText /> matches snapshot 3`] = `
<Link
  to="/disambiguation/1,2"
>
  Link
</Link>
`;
